packages/astro/src/vite-plugin-astro/index.ts

```diff
@@ -43,35 +41,57 @@ export default function astro({ settings, logger }: AstroPluginOptions): vite.Pl
 	const prePlugin: vite.Plugin = {
 		name: 'astro:build',
 		enforce: 'pre', // run transforms before other plugins can
-		configResolved(_resolvedConfig) {
-			resolvedConfig = _resolvedConfig;
+		configResolved(viteConfig) {
+			// Initialize `compile` function to simplify usage later
+			compile = (code, filename) => {
+				return compileAstro({
+					compileProps: {
+						astroConfig: config,
+						viteConfig,
+						preferences: settings.preferences,
+						filename,
+						source: code,
+					},
+					astroFileToCompileMetadata,
+					logger,
+				});
+			};
 		},
 		configureServer(_server) {
 			server = _server;
+			// Make sure deleted files are removed from the compile metadata to save memory
+			server.watcher.on('unlink', (filename) => {
+				astroFileToCompileMetadata.delete(filename);
+			});
 		},
 		buildStart() {
 			astroFileToCssAstroDeps = new Map();
+			astroFileToCompileMetadata = new Map();
+
+			// Share the `astroFileToCompileMetadata` across the same Astro config as Astro performs
+			// multiple builds and its hoisted scripts analyzer requires the compile metadata from
+			// previous builds. Ideally this should not be needed when we refactor hoisted scripts analysis.
+			if (astroFileToCompileMetadataWeakMap.has(config)) {
+				astroFileToCompileMetadata = astroFileToCompileMetadataWeakMap.get(config)!;
```

Comment:
I'm not really sure about `@ts-expect-error` since we only need to suppress a specific error instead of the whole line. I think once TS improves this, our eslint config should flag it: 
<img width="644" alt="image" src="https://github.com/withastro/astro/assets/34116392/6b289dc4-134d-47c8-92d9-4cfa59ae33d3">